hudi-common/src/main/java/org/apache/hudi/metadata/FileSystemBackedTableMetadata.java

```diff
@@ -96,11 +109,32 @@ public List<String> getPartitionPathWithPathPrefixes(List<String> relativePathPr
   }
 
   private List<String> getPartitionPathWithPathPrefix(String relativePathPrefix) throws IOException {
+    return getPartitionPathWithPathPrefixUsingFilterExpression(relativePathPrefix, null, null);
+  }
+
+  private List<String> getPartitionPathWithPathPrefixUsingFilterExpression(String relativePathPrefix,
+                                                                           Types.RecordType partitionFields,
+                                                                           Expression expression) throws IOException {
     List<Path> pathsToList = new CopyOnWriteArrayList<>();
     pathsToList.add(StringUtils.isNullOrEmpty(relativePathPrefix)
-        ? new Path(datasetBasePath) : new Path(datasetBasePath, relativePathPrefix));
+        ? dataBasePath.get() : new Path(dataBasePath.get(), relativePathPrefix));
     List<String> partitionPaths = new CopyOnWriteArrayList<>();
 
+    int partitionLevel = -1;
+    boolean needPushDownExpressions;
+    // Not like `HoodieBackedTableMetadata`, since we don't know the exact partition levels here,
+    // given it's possible that partition values contains `/`, which could affect
+    // the result to get right `partitionValue` when listing paths, here we have
+    // to make it more strict that `urlEncodePartitioningEnabled` must be enabled.
+    // TODO better enable urlEncodePartitioningEnabled if hiveStylePartitioningEnabled is enabled?
```

Comment:
We have to list at least one partition path to get the partition levels for `FileSystemBackedTableMetadata` to ensure the number of partition columns and partition levels are same. But it might be time consuming especially for object store.

So here I have to make it more strict for `FileSystemBackedTableMetadata` to check `urlEncodePartitioningEnabled` also.  While we don't need to check it in `HoodieTableBackedTableMetadata`, since we can simply get one partition path to check it.

I'm thinking maybe we should enable `urlEncodePartitioningEnabled` if `hiveStylePartitioningEnabled` is enabled? Since this is the default behavior for spark and hive

Reply:
Shall we disable partition push down filters by default when FileSystemBackedTableMetadata is used ?

Reply:
Checking hudi configure here is difficult, since`HoodieConfig` is not passed when construct `FileSystemBackedTableMetadata`, that's why this class lefts some TODOs here like `TODO: Get the parallelism from HoodieWriteConfig` and `get `caseSensitive` from HoodieConfig`.

I think it's safe if we both check `hiveStylePartitioningEnabled` and `urlEncodePartitioningEnabled` when using `FileSystemBackedTableMetadata` and enable partition push down filters.

And I'll try to resolve TODOs relate to HoodieConfig in a follow-up pr.

Reply:
@boneanxs : Made one more pass at it. Yes, Lets keep partition push down filters disabled if urlEncodePartitioningEnabled is disabled. Once you address all comments, let me know. Thanks.